drivers: flash_stm32_ospi: enable basic support for PM

bearsh opened the pull request:
also add a workaround for an errata which leads to corrupted data on the first read after exiting stop 2 or 3

bearsh committed:
drivers: flash_stm32_ospi: enable basic support for PM

lock pm state and device while doing transfers

Signed-off-by: Martin Gysel <me@bearsh.org>

bearsh committed:
drivers: flash_stm32_ospi: workaround for corrupted data

this is a workaround for an errata which leads to corrupted data on the
first read after exiting stop 2 or 3

Signed-off-by: Martin Gysel <me@bearsh.org>

erwango commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -24,3 +24,27 @@ config FLASH_STM32_OSPI
+
+if FLASH_STM32_OSPI
+
+if SOC_STM32U595XX || SOC_STM32U595XX || SOC_STM32U595XX || SOC_STM32U595XX \
+	|| SOC_STM32U595XX || SOC_STM32U595XX || SOC_STM32U595XX || SOC_STM32U595XX \
+	|| SOC_STM32U595XX || SOC_STM32U595XX || SOC_STM32U599XX || SOC_STM32U599XX \
+	|| SOC_STM32U599XX || SOC_STM32U599XX || SOC_STM32U599XX || SOC_STM32U599XX \
+	|| SOC_STM32U599XX || SOC_STM32U5A5XX || SOC_STM32U5A5XX || SOC_STM32U5A5XX \
+	|| SOC_STM32U5A5XX || SOC_STM32U5A5XX || SOC_STM32U5A9XX || SOC_STM32U5A9XX \
+	|| SOC_STM32U5A9XX || SOC_STM32U5A9XX
```
can you review this list ?

bearsh replied:
grepping trough the zephyr source gives me following list:
```
SOC_STM32U595XX
SOC_STM32U599XX
SOC_STM32U5A5XX
SOC_STM32U5A9XX
```
should I limit it to SOCs mentioned in the source tree?

erwango commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -24,3 +24,27 @@ config FLASH_STM32_OSPI
+config FLASH_STM32U5_ES0553_2_5_13
+	bool "Workaround for errata 2.5.13 of ES0553"
+	default y if PM
+	default n if !PM
```
This line shouldn't be required

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -226,6 +229,18 @@ static int ospi_read_access(const struct device *dev, OSPI_RegularCmdTypeDef *cm
+
+	if (dev_data->errata_u59_2_5_13) {
+		dev_data->errata_u59_2_5_13 = false;
+		err = ospi_read_access(dev, cmd, data, 1);
```
@FRASTM can you confirm this is fine ?

bearsh replied:
according to the errata sheet, it is only needed after exiting stop 2 or 3. but then we need a way to get notified about that, which AFAIK may be possible with PM hooks. should we go that way?

mathieuchopstm commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -24,3 +24,27 @@ config FLASH_STM32_OSPI
+	default y if PM
+	default n if !PM
+	help
+	  Handles erratum "OCTOSPI external memory read issue after exiting
+	  Stop 2 or Stop 3 mode when using DQS and delay block."
+	  Seen in Errata Sheet ES0553 §2.5.13 for STM32U5x5/STM32U5x9.
```
Proposal:
```suggestion
	  Enable workaround for erratum "OCTOSPI external memory
	  read issue after exiting Stop 2 or Stop 3 mode when using
	  DQS and delay block" documented in section 2.5.13 of errata
	  sheet ES0553 revision 5. 
```

mathieuchopstm commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -222,6 +229,18 @@ static int ospi_read_access(const struct device *dev, OSPI_RegularCmdTypeDef *cm
+#if CONFIG_FLASH_STM32U5_ES0553_2_5_13
+	int err;
+
+	if (dev_data->errata_u59_2_5_13) {
+		dev_data->errata_u59_2_5_13 = false;
+		err = ospi_read_access(dev, cmd, data, 1);
+		if (err) {
+			return err;
+		}
+	}
```
Can't we just trigger a dummy read straight from the PM callback?

mathieuchopstm replied:
> > Can't we just trigger a dummy read straight from the PM callback?
> 
> just that we talk about the same thing. Your idea was to do the read in `flash_stm32_ospi_pm_action()`->`flash_stm32_ospi_activate()`?

What I meant is performing the call: `err = ospi_read_access(dev, cmd, data, 1);`

But this is a very naive question because I'm not familiar with this driver - it may very well not be possible, in which case ignore the remark.

> > The user is going to read at some point, ...
> 
> not necessarily. if you intent to simply store data **for a certain time**, there's no need to read before write after each wakeup.

Emphasis mine. But let's not fight over this.
I don't mind (too much) what you're proposing, I'm merely trying to make the proposal better with my understanding and vision on things (which can be wrong!).

If there is a way to implement what I proposed earlier:
> [...] start an asynchronous read whose result is discarded [from the PM action]

then I believe the errata handling should be done this way. Otherwise, keep as-is and close this point.

(actually, I have a Nit about current implementation, but that's a small detail)

mathieuchopstm commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -24,3 +24,27 @@ config FLASH_STM32_OSPI
+	|| SOC_STM32U5A5XX || SOC_STM32U5A5XX || SOC_STM32U5A9XX || SOC_STM32U5A9XX \
+	|| SOC_STM32U5A9XX || SOC_STM32U5A9XX
+
+config FLASH_STM32U5_ES0553_2_5_13
```
Nit: a more descriptive name would appreciated, e.g.:
```suggestion
config FLASH_STM32_ERRATUM_U5_STOP2_3_CORRUPT_READ
```

mathieuchopstm commented on drivers/flash/Kconfig.stm32_ospi:
```diff
@@ -24,3 +24,27 @@ config FLASH_STM32_OSPI
+	|| SOC_STM32U5A9XX || SOC_STM32U5A9XX
+
+config FLASH_STM32U5_ES0553_2_5_13
+	bool "Workaround for errata 2.5.13 of ES0553"
```
Nit:
```suggestion
	bool "Workaround for corrupted reads after Stop 2/3 exit"
```

bearsh commented:
I'm wondering if it make sense / is desirable if I pinctrl handling to pm action handler? Any opinion?

bearsh commented:
Would you also accept an option to disable ospi auto polling but use software polling. depending on write/erase speed of the flash, I quite sure, it's more energy efficient to do sw polling where it is possible to go to sleep in between.

bearsh committed:
drivers: flash_stm32_ospi: apply pinctrl states in device's pm actions

Apply pinctrl sleep, if available, on suspend, restore default on
resume.
Rename resume/suspend functions to better express their meaning.

Signed-off-by: Martin Gysel <me@bearsh.org>

mathieuchopstm commented on drivers/flash/flash_stm32_ospi.c:
Considering `ospi_pm_get` (resp. `ospi_pm_put`) seems to always follow an `ospi_thread_lock` (resp. `ospi_thread_unlock`), it would be more sensible to do both at the same time in a single function.

mathieuchopstm commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -2708,6 +2723,9 @@ static int flash_stm32_ospi_pm_action(const struct device *dev, enum pm_device_a
 	case PM_DEVICE_ACTION_RESUME:
+#if CONFIG_FLASH_STM32_ERRATUM_U5_STOP2_3_CORRUPT_READ
+		dev_data->post_wakeup_dummy_read_needed = true;
+#endif
```
I think setting the flag in `PM_DEVICE_ACTION_SUSPEND` would be more sensible

bearsh replied:
done

bearsh committed:
drivers: flash_stm32_ospi: add helper for thread locking an pm_get/put

as ospi_lock_thread()/ospi_pm_get() and
ospi_pm_put()/ospi_unlock_thread() are always called together, combine
them in a helper function.

Signed-off-by: Martin Gysel <me@bearsh.org>

bearsh commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -2195,6 +2231,49 @@ static int spi_nor_process_bfp(const struct device *dev,
 
+#ifdef CONFIG_PM_DEVICE
+
+static int flash_stm32_ospi_suspend(const struct device *dev)
```
done

mathieuchopstm reviewed (commented):
Seems fair in current state. I'll let @FRASTM and @erwango review and give +1

mathieuchopstm commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -2186,6 +2236,64 @@ static int spi_nor_process_bfp(const struct device *dev,
+
+#endif
+
+static int ospi_resume(const struct device *dev)
```
That would be fine by me.

mathieuchopstm reviewed (commented):
One thing I had seen but didn't check until now

mathieuchopstm reviewed (commented):
LGTM. Leaving +1 to @FRASTM and @erwango

bearsh commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -2588,9 +2654,32 @@ static int flash_stm32_ospi_init(const struct device *dev)
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		err = flash_stm32_ospi_activate(dev);
```
I see, it seems I messed up when rebasing/amending my commits... 

but is it ok the change the name of the function by a follow-up commit as requested by @mathieuchopstm in https://github.com/zephyrproject-rtos/zephyr/pull/94939#discussion_r2619176077?
